### PR TITLE
Decouple Forward and Backward of ATenOp

### PR DESCRIPTION
**Pull request by iK1D**
Decouple forward and backward for ATenOp, so the ATenOp forward (for inference or node doesn't require grad) can run. The PR includes:
- Remove attribute for ATenOp, all arguments will be inputs of ATenOp node, those non-tensor arguments will be CPU inputs.
- Use op_name and overload_name to determin the ATen operator.
- Remove ATenOp config for ATenOp forward run. All inputs will be converted to DLPacks, torch extension will convert these DLPacks to ATen operator arugments according to the operator schema and run.
- Remove ATenOpGrad. Use ATenOp node for both forward and backward.
- Move ATenOp's gradient definition to frontend. Use Python code to define ATenOp's gradient, we can use multiple ONNX Ops and ATenOp nodes to define a single ATenOp's gradient. The definition will be passed to backend in JSON format when backend build the gradient graph for ATenOp nodes.
- Move PyTorch exporter overriding function to the same place as the ATenOp gradient definitions, and also use Python decorator to register them.
- Add support to torch.argmax for forward run only, which is required by one of the models we are investigating on for now. ORT currently doesn't have CUDA kernel for ArgMax for OpSet12 or larger, because we are using CUDNN for the implementation for ArgMax, but CUDNN cannot support the new attribute "select_last_index" added since OpSet 12.

Moving forward:
- This change can support inference for those graph contains ATenOp node generated by exporter. Currently all codes are under ENABLE_TRAINING, in the future to enable running for inference, we need to move ATenOp schema and kernel from orttraining folder to onnxruntime folder.

### CODE REVIEW

**Review by thiagocrepaldi — APPROVED** [review]

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/__init__.py [hunk 1]
```diff
@@ -13,6 +13,11 @@
 import threading
 from functools import wraps
 from onnxruntime.capi import _pybind_state as C
+from torch.onnx import register_custom_op_symbolic
+from ._aten_op_factory import ATenOpFactory
+
+
+_onnx_opset_version = 1
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/__init__.py:20]: This information is available at https://github.com/microsoft/onnxruntime/blob/master/orttraining/orttraining/python/training/ortmodule/__init__.py#L16

In fact, global configuration should be in a single central spot as opposed to spread throughout the code

Why `1` and not `12`?

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 2]
```diff
@@ -224,7 +224,7 @@ def _export_model(self, *inputs, **kwargs):
 
         self._set_device_from_module(inputs, kwargs)
         self._onnx_model = self._get_exported_model(*inputs, **kwargs)
-        _cpp_ext._load_aten_op_executor_cpp_extension_if_needed(self._onnx_model, self._loglevel < _logger.LogLevel.WARNING)
+        _cpp_ext._load_aten_op_executor_cpp_extension_if_needed(self._onnx_model)
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:227]: Will this make the backend more verbose
? This flag was used to suppress unnecessary logs in production

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+"""Support registration of ATenOp's symbolic (PyTorch exporter overriding) and gradient definition.
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py:6]: If docstring is used, this will be published as part of ORTModule documentation. If this is implementation detail, better to replace it with `# regular comments` as opposed to `'''docstring '''`

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
+# -------------------------------------------------------------------------
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py:1]: This file is not a PyTorch CPP extension, as described here https://pytorch.org/docs/stable/cpp_extension.html

I would prefer moving this file outside torch_cpp_extensions as it can interfere with onnxruntime packaging

maybe `orttraining/orttraining/python/training/ortmodule/_aten_op_symbolics_factory.py`?

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                  name_s='aten::embedding')
+    indices_shape = _get_tensor_sizes(indices)
+    if indices_shape is not None and hasattr(weight.type(), 'with_sizes'):
+        output_type = weight.type().with_sizes(
+            indices_shape + [_get_tensor_dim_size(weight, 1)])
+        output.setType(output_type)
+    return output
+
+
+@ATenOpFactory.register_gradient('aten::embedding')
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py:83]: this is really cool...
I think register_gradient method should have it's own class, as this can also be used for non-ATenOp

File: orttraining/orttraining/core/graph/aten_op_gradient.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+  std::string op_type;
+  std::string domain = kOnnxDomain;
+  std::vector<std::string> inputs;
+  std::vector<std::string> outputs;
+  std::unordered_map<std::string, AttributeDefinition> attributes;
+};
+
+void ParseATenOpGradientDefinition(const std::string& grad_def_json_str, std::vector<NodeDefinition>& grad_def);
+
+class ATenOpGradientDefinitionGetter {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/aten_op_gradient.h:44]: For this PR, it's ok if it only works for ATenOp. 
But we can extend this approach for defining gradients for any ops.

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                    output of forward node.
+                'b' is True or False, indicating if this attribute is tensor_proto ot not.
+                    'is_tensor' is optional, if not present, the default is False.
+"""
+
+import json
+from torch.onnx.symbolic_helper import _get_tensor_dim_size, _get_tensor_sizes
+
+
+def gradient_definition_to_json(gradient_def):
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py:32]: I know json is an easy bridge between python structure-free strings to backend's parser, but I am wondering if we should introduce some structure to represent this gradient definition.

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/__init__.py (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/__init__.py:20]: Just follow here to https://github.com/microsoft/onnxruntime/blob/bcd50afafbb95d9206535802669a7b4a9ec5a019/tools/python/register_custom_ops_pytorch_exporter.py#L11 to use 1 instead of other version.

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py (hunk 2 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:227]: I remove it because he parameter is not used in the function.

File: orttraining/orttraining/core/graph/aten_op_gradient.cc [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+#include "orttraining/core/graph/aten_op_gradient.h"
+
+#include <nlohmann/json.hpp>
+
+namespace onnxruntime {
+namespace training {
+
+using json = nlohmann::json;
+
+void ParseATenOpGradientDefinition(const std::string& grad_def_json_str, std::vector<NodeDefinition>& grad_def) {
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/aten_op_gradient.cc:13]: Something to consider, if we directly pybind NodeDefinition and AttributeDefinition, we won't need json serialization and parsing. 
Not a blocking comment for this PR, but let's think about the most intuitive and expressive way to define gradient in python.

File: orttraining/orttraining/core/graph/training_op_defs.cc [hunk 9]
```diff
@@ -2805,53 +2801,20 @@ Return true if all elements are true and false otherwise.
               /*is_homogeneous*/ false,
               /*min_arity*/ 1)
       .Attr("name", "Name of ATen operator.", AttributeProto::STRING)
-      .Attr("custom_attributes_json", "custom attributes in JSON format.", AttributeProto::STRING, false)
+      .Attr("overload_name", "Overload name of ATen operator.", AttributeProto::STRING, false)
+      .Attr("output_types", "Output types of ATenOp's gradient op.", AttributeProto::INTS, false)
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/training_op_defs.cc:2805]: is "output_types" specific to gradient op only? 
How do we set output type for ATen forward ops?

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py (hunk 5 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py:83]: However, this gradient builder doesn't have reference to the forward node. 
What if we wish to express some complex gradient building logic? e.g. check if broadcast handling is needed, check num_inputs in original node....

File: orttraining/orttraining/core/graph/gradient_builder.cc [hunk 10]
```diff
@@ -1683,41 +1682,68 @@ IMPLEMENT_GRADIENT_BUILDER(GetMinMaxGradient) {
...
-  std::vector<ArgDef> input_args;
-  std::vector<ArgDef> output_args;
+    std::vector<ArgDef> output_args;
+    std::vector<int64_t> grad_output_types;
+    for (const auto& output : node_def.outputs) {
+      if (output.find("GI(") == 0) {
+        size_t index = static_cast<size_t>(std::stoi(output.substr(3, output.length() - 4)));
+        output_args.emplace_back(GI(index));
+        if (is_aten_op) {
+          grad_output_types.emplace_back(IElemType(index));
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/graph/gradient_builder.cc:1723]: in addition to type, we can also pass shape here.

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc [hunk 11]
```diff
@@ -6,11 +6,107 @@
...
+    TORCH_INTERNAL_ASSERT(index < argument_size);
+    bool is_optional = is_optional_arguments[index];
+    TORCH_INTERNAL_ASSERT(dlpack || is_optional || default_values[index]);
+    if (!dlpack) {
+      if (is_optional) {
+        // Optional argument always has no default value.
+        return c10::IValue(c10::nullopt);
+      }
+
+      return *default_values[index];
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:66]: I didn't check c10:optional class signature, but if the "*" is expected?

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc [hunk 12]
```diff
@@ -20,90 +116,83 @@ class ATenOperatorCache {
...
+
+  const char* GetGradientDefinition(const std::string& op_name, const std::string& overload_name) {
+    auto it = gradient_definitions_.find(std::make_pair(op_name, overload_name));
+    return it != gradient_definitions_.end() ? it->second.c_str() : nullptr;
   }
 
  private:
   ATenOperatorCache() = default;
-  std::unordered_map<std::string, ATenOperator> ops_;
+  std::unordered_map<std::pair<std::string, std::string>, ATenOperator, PairHash> ops_;
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:171]: looks like worth defining a struct ATenOpType?

**Comment by SherlockNoMad** [comment]
In principle, I am ok with merging this PR for this iteration. 
I think we need to rethink about how we compose Gradient Definition, and how to pass it from python to c++.

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py (hunk 7 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py]: Previously I though I needed the torch extension to pass by the definition, then string is the best as we didn't need to define structure on both ORT and torch extension. But on second thought, we don't need the torch extension for this. I've changed to use pybind for this.

File: orttraining/orttraining/core/graph/aten_op_gradient.cc (hunk 8 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/aten_op_gradient.cc]: Changed to use pybind.

File: orttraining/orttraining/core/graph/training_op_defs.cc (hunk 9 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/training_op_defs.cc:2805]: This is the issue I mentioned before, that we need to add some code in symbolic_shape_infer.py to support this. Ideally we need exporter to help, by setting some information in the node attribute to help us.

File: orttraining/orttraining/core/graph/gradient_builder.cc (hunk 10 above)
Comment (iK1D) [reply orttraining/orttraining/core/graph/gradient_builder.cc:1716]: List of list (may have different sizes) is not easy to support in node attribute...

File: orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc (hunk 11 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/aten_op_executor/aten_op_executor.cc:66]: PyTorch code also uses in such way.

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc [hunk 13]
```diff
@@ -308,7 +308,30 @@ Status GradientGraphBuilder::Build(const std::unordered_set<std::string>* p_init
     }
   }
 
-  return GraphAugmenter::AugmentGraph(*graph_, gradient_graph_defs, p_initializer_names_to_preserve);
+  ORT_RETURN_IF_ERROR(GraphAugmenter::AugmentGraph(*graph_, gradient_graph_defs, p_initializer_names_to_preserve));
+
+  // For ORTModule, the graph will be sent back to frontend for cache. Frontend will use inference session to run the
+  // cached graph. We need to save the new NodeArgs to value_info so we will not loss those shape/type information
+  // during graph to graph_proto conversion. Skip the graph outputs as they are already in the output section in
```
Comment (SherlockNoMad) [inline orttraining/orttraining/core/framework/gradient_graph_builder.cc:315]: From the AddValueInfo impl, duplicated insert will error out: "Error: trying to add an existing value info." 

FYI, AddValueInfo has a really bad impl, where it iterate through the value_info to check duplicate. 
We should use std::unordered_set for value_info_

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        output_type = weight.type().with_sizes(
+            indices_shape + [_get_tensor_dim_size(weight, 1)])
+        output.setType(output_type)
+    return output
+
+
+@register_symbolic('max_pool2d')
+def max_pool2d(g, self, kernel_size, stride, padding, dilation, ceil_mode):
+    return g.op("com.microsoft::ATenOp", self, kernel_size, stride, padding, dilation, ceil_mode,
+                name_s='aten::max_pool2d_with_indices', outputs=2)[0]
```
Comment (SherlockNoMad) [inline orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:75]: Just checking, do we need [0] here?

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc (hunk 13 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/core/framework/gradient_graph_builder.cc:315]: with populated value_info, we should start seeing shape information in saved *_training.onnx and *_optimized_training.onnx graph. You can please check to verify.

**Review by SherlockNoMad — APPROVED** [review]
Good to merge after fixing the AddValueInfo function.

File: orttraining/orttraining/core/framework/gradient_graph_builder.cc (hunk 13 above)
Comment (iK1D) [reply orttraining/orttraining/core/framework/gradient_graph_builder.cc:315]: Yes, I can see these information in the graph after the fix.

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py (hunk 14 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:75]: Yes we need that. The return needs to match the max_pool2d (Python func) in PyTorch's side, which has only one return value. The ATenOp (C++ func) returns 2 tensors. When removing [0], the exporter will fail.

**Comment by iK1D** [comment]
> Good to merge after fixing the AddValueInfo function.

I will create another PR to change the AddValueInfo function as this is an existing one and not related to this PR.
